### DISCUSSION TITLE
GM861 Barcode/QR reader not sending data

### DISCUSSION THREAD

**Discussion by DutchCoban** [arendst/Tasmota#18399] [Support request] [answered]
Hi there,

I've purchased a [GM861](http://www.zjgrow.com/grow-gm861-uartusb-interface-1d2d-bar-code-qr-code-barcode-scanner-module-with-stainless-steel-enclosure-white-green-led-p4316421.html) barcode/QR scanner that I've hooked up on a Wemos D1 mini pro the following way:
- GM861 RX is connected to D5 and that I've set to "Serial TX",
- GM861 TX is connected to D6 and that I've set to "Serial RX",
- GM861 GND is connected to GND,
- GM861 3.3 is connected to 3.3v,
- Baudrate is set to 9600.

Whenever I scan a barcode the ring on the GM861 turns green, meaning that it has read the barcode succesfully. Only thing is that that code doesn't show up in the console. I've used the "classic" Tasmota and Tasmota Sensors, but with my settings it doesn't show a code. I've looked at a post on [Arduino.cc](https://forum.arduino.cc/t/how-to-use-gm61-qr-code-scanner/644030/3) where someone wrote an .INO to publish the bar/QR-codes. How do I get Tasmota to show (and later on publish) the barcode, just like with a regular meterreading of a sensor?

For those wondering: I was thinking of implementing this in my kitchen so I can read the barcodes of things we need to buy. Send the barcode to Home Assistant, create an automation that puts it in my shopping list.

**Comment by barbudor** [comment]
SerialTX/SerialRX will not work on GPIO ohters than the native ESP8266 UART GPIO 1 & 3 - but in that case you must disconnect USB from your Wemos and power with 5V through the headers.
You can try with the software UART SerialBridge by using SerBrTX and SerBrRX (commands are slighly different, so chack the doc)

When having to deal with serial devices, it's more reliable to go for a ESP32 or ESP32-C3 which have more hardware UARTs.

**Comment by DutchCoban** [comment]
I don't have an ESP32-board at hand (at the moment). I've tried setting up the SerialBridge but I couldn't get it to send the data. I've found the datasheet/manual and this would be the most interesting part I presume: Baudrate 9600 bps; No check; 8 bit data; 1 bit stopbits; No flowcontrol. I've set it all, but no data.

https://www.dropbox.com/sh/buysgr2aeutuub3/AADsvupiGseFm-6IowMBWGCOa?dl=0&preview=GM861+Barcode+reader+module+User+Manual.pdf

**Comment by FukushimaShouten** [comment]
Hi Crispy78,

If I understand your problem correctly, you get no scanned data transmitted from GM861's TXD output, right?
If so, I got the same problem. Just like you, the module's led ring turns green when there is a QR code in it's range but no sign of serial activity on TXD.

I just gave some  random tries and got at least something by following the steps below

1. Power up the module
2. Scan the 'Default: setup code on' QR code on the datasheet <<1.6 Setup code>>
3. Scan the 'Output' QR code on the same page
4. Same drill, 'Serial output' on <<2.1 Serial Communication Interface>>

After that I got something (6..7 digit numbers?) came out from the module.

then I scanned the QR codes on <<1.1.1 Read Version Information>> and <<1.1.2 Read the Unique Device ID>>, both gave me seemingly correct (meaningful) strings shown on my terminal software.

My set up is as follows:-

GM861 - Serial/USB dongle - PC running teraterm, config'd 9600-8-N-1, no flow control

Wireing:- 
Red - Green : external power supply 3.3V - GND.
Black: dongle RXD
Yellow: dongle TXD
and of course dongle GND is connected to the power supply GND.

As you may have already found out, the datasheet is horribly incomprehensible so I ended up with guessing game but at least i made a bit of headway so hopefully this also helps you.

**Reply by DutchCoban** [reply]
When I connect it to my computer using your suggestions, it's working. Now I need it to work with a ESP8266 and Tasmota :-)

**Comment by FukushimaShouten** [comment]
Just to give you a quick update. I also scanned couple of QR codes printed on a paper and the module gave me the URLs burried in the codes seemingly correct.

**Reply by DutchCoban** [reply]
I haven’t scanned any code, that might be the problem. I’m going to fiddle with it in one of the upcoming days.

**Comment by arendst** [comment]
I ordered one. Takes three weeks to arrive. I'll see if I can write a driver for it.

**Reply by DutchCoban** [reply]
It's a nice sensor, don't forget to use the manual (https://github.com/arendst/Tasmota/discussions/18399#discussioncomment-5608039)  and @FukushimaShouten's advice when setting up :-) The sensor outputs the read codes in plaintext.

**Comment by DutchCoban** [comment]
EUREKA!!! :-)

I've used a Wemos D1 loaded with http://ota.tasmota.com/tasmota/tasmota.bin.gz.
I've setup the GM861 using @FukushimaShouten advise on https://github.com/arendst/Tasmota/discussions/18399#discussioncomment-5733038 and entered `Backlog SerialLog 2; Baudrate 9600; SerialConfig 8N1; SerialSend 2` in Tasmota (borrowed from: https://groups.google.com/g/sonoffusers/c/uY4vLoZBZP0). 

My console output is now:
```
18:01:00.318 CMD: Backlog SerialLog 2; Baudrate 9600; SerialConfig 8N1; SerialSend 2
18:01:00.724 RSL: RESULT = {"Command":"Unknown"}
18:01:00.961 SER: Set to 8N1 9600 bit/s
18:01:00.964 RSL: RESULT = {"Baudrate":9600}
18:01:01.215 RSL: RESULT = {"SerialConfig":"8N1"}
18:01:01.464 RSL: RESULT = {"SerialSend":"Done"}
18:01:02.019 RSL: RESULT = {"SerialReceived":"8712800188339\r"}
18:01:03.067 RSL: RESULT = {"SerialReceived":"8712800188339\r"}
18:01:04.023 RSL: RESULT = {"SerialReceived":"8712800188339\r"}
18:01:05.074 RSL: RESULT = {"SerialReceived":"8712800188339\r"}
18:01:06.028 RSL: RESULT = {"SerialReceived":"8712800188339\r"}
18:01:07.030 RSL: RESULT = {"SerialReceived":"8712800188339\r"}
```

**Reply by barbudor** [reply]
Enjoy your Chocomel @crispy78 🤣

**Comment by FukushimaShouten** [comment]
Good to know it worked on your side as well..! Thanks for updating me with your result :)

**Comment by DutchCoban** [comment]
Any idea's how I can add the GM861 data as a sensor to my Tasmota device in Home Assistant?

I was hoping it was as easy as "rule2 ON SerialReceived#Data DO Publish stat/%topic%/SENSOR {"Barcode":%value%} ENDON", I've been Googleling but I might have overlooked a thing or two or I'm thinking way to easy or way to difficult.

It used to publish these in the logs:
18:01:05.074 RSL: RESULT = {"SerialReceived":"8712800188339\r"}
18:01:06.028 RSL: RESULT = {"SerialReceived":"8712800188339\r"}
18:01:07.030 RSL: RESULT = {"SerialReceived":"8712800188339\r"}

I've got it to publish these via MQTT (don't ask me how 😃)
21:58:32.881 MQT: tele/35/RESULT = {"SerialReceived":"8712800188339\r"}
21:58:33.830 MQT: tele/35/RESULT = {"SerialReceived":"8712800188339\r"}
21:58:34.843 MQT: tele/35/RESULT = {"SerialReceived":"8712800188339\r"}

So I could build a Home Assistant MQTT sensor, it would be nicer if it displayed the data as a sensor on the device page of HA.

**Reply by barbudor** [reply]
I'm not using HA so my word is far absolute accuracy but I understand there are 2 ways
1) manual way by creating a YML file which describe the sensor. It could probably work even with the default serial received message. Don't ask me more details on how to write that file
2) auto discovery if you have the hasmota plug-in
By default on start tasmota sends a discovery message on a specific topic. There are 2 messages actually
- 1st give some details on the Tasmota unit
- 2nd describe the SENSOR message and it's nothing more than a SENSOR message but on this specific discovery topic.
I don't know by heart those discovery topic but you can set Weblog 4 and restart. You will see tasmota publishing those (they start by tasmota/discovery)
Or if you have a MQTT monitoring tool, looks at what is publish under tasmota/discovery

Once you know the exact topic send your SENSOR message once on that topic with Publish2 (to make it retained) but after the official tasmota one
Something like
On mqtt#connected do RuleTimer1 12 end on
On rules#Timer=1 do Publish2 tasmota/discovery/.... Endon
The SENSOR message here just need to be dummy, manually made. Just to show HA the structure

Then your rule as above should be handled by HA

**Comment by arendst** [comment, marked as answer]
Try latest dev branch (v12.5.0.2) and enable GM861 driver with ``#define USE_GM861``.

- Select GPIOs and start scanning! No need to scan the init sequence from the User Guide.
- If it fails try to reset and re-init the scanner with command ``Gm861Reset 1``
- If needed configure zone bytes using command ``Gm861Zone<byte> <value>``  like ``Gm861Zone0``.
- Dump zone bytes by enabling logging level 4 and execute command ``Gm861Dump``.

```00:00:00.001 HDW: ESP8266EX
00:00:00.046-032 CFG: Loaded from flash at FB, Count 701
00:00:00.129-032 QPC: Reset
00:00:00.140-032 ROT: Mode 1
00:00:00.141-032 SRC: Restart
00:00:00.144-032 Project tasmota - Wemos4 Version 12.5.0.2(theo)-2_7_4_9(2023-05-05T16:58:53)
00:00:00.447-032 WIF: Attempting connection...
00:00:00.784-032 WIF: Connecting to AP1 indebuurt_IoT Channel 11 BSSId 18:E8:29:CA:17:C1 in mode 11n as wemos4...
00:00:01.754-032 WIF: Connected
00:00:02.006-029 HTP: Web server active on wemos4 with IP address 192.168.2.223
00:00:02.412-029 WIF: DNS resolved 'loghost' (192.168.2.24) in 8 ms
00:00:02.429-029 WIF: DNS resolved 'ntpserver1' (192.168.2.27) in 4 ms
00:00:02.442-029 RTC: UTC 2023-05-05T14:59:52, DST 2023-03-26T02:00:00, STD 2023-10-29T03:00:00
16:59:52.000-029 RTC: Synced by NTP
16:59:53.005-029 MQT: Attempting connection...
16:59:53.011-029 WIF: DNS resolved 'mqttbroker' (192.168.2.24) in 4 ms
16:59:53.020-029 MQT: Connected
16:59:53.023-026 MQT: tele/wemos4/LWT = Online (retained)
16:59:53.024-026 MQT: Subscribe to cmnd/wemos4/#
16:59:53.028-027 MQT: Subscribe to cmnd/tasmotas/#
16:59:53.032-027 MQT: Subscribe to cmnd/theotest1/#
16:59:53.033-026 MQT: Subscribe to cmnd/DVES_83BB10_fb/#
16:59:53.042-027 MQT: tele/wemos4/INFO1 = {"Info1":{"Module":"Generic","Version":"12.5.0.2(theo)","FallbackTopic":"cmnd/DVES_83BB10_fb/","GroupTopic":"cmnd/tasmotas/"}}
16:59:53.056-027 MQT: tele/wemos4/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"wemos4","IPAddress":"192.168.2.223"}}
16:59:53.064-025 MQT: tele/wemos4/INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":154}}
16:59:54.285-029 GM8: Initialized
16:59:57.392-025 MQT: tele/wemos4/STATE = {"Time":"2023-05-05T16:59:57","Uptime":"0T00:00:09","UptimeSec":9,"Heap":28,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":20,"MqttCount":1,"Wifi":{"AP":1,"SSId":"indebuurt_IoT","BSSId":"18:E8:29:CA:17:C1","Channel":11,"Mode":"11n","RSSI":100,"Signal":-34,"LinkCount":1,"Downtime":"0T00:00:03"}}
16:59:58.390-028 APP: Boot Count 154
16:59:58.647-028 CFG: Saved to flash at FA, Count 702, Bytes 4096
17:00:05.003-024 MQT: tele/wemos4/SENSOR = {"Time":"2023-05-05T17:00:04","GM861":"8711218972875"}
17:02:01.214-027 CMD: seriallog 4
17:02:01.216-027 SRC: Serial
17:02:01.218-027 CMD: Grp 0, Cmd 'SERIALLOG', Idx 1, Len 1, Pld 4, Data '4'
17:02:01.226-024 MQT: stat/wemos4/RESULT = {"SerialLog":{"4":{"Active":"4"}}}
17:02:01.622-027 CFG: Saved to flash at F9, Count 703, Bytes 4096
17:02:07.078-027 CMD: gm861dump
17:02:07.079-027 SRC: Serial
17:02:07.082-027 CMD: Grp 0, Cmd 'GM861DUMP', Idx 1, Len 0, Pld -99, Data ''
17:02:07.085-027 DMP: 7E 00 07 01 00 00 62 5D 84
17:02:07.101-024 MQT: stat/wemos4/RESULT = {"GM861Dump":"Done"}
17:02:07.227-027 DMP: 02 00 00 62 D6 00 20 01 00 0A 32 01 2C 00 87 3C 01 A0 1C 32 03 00 80 00 06 00 00 00 00 00 00 00 00 00 00 00 00 02 80 3C 00 00 00 06 00 00 39 01 05 64 0D 0D 0D 01 0D 01 04 80 09 04 80 05 04 80 01 04 80 01 08 04 80 08 04 80 08 04 80 08 04 80 08 01 80 00 00 00 04 80 01 01 00 00 00 00 04 80 00 00 03 00 01 00 2D 9E
17:02:10.634-027 CMD: seriallog 3
17:02:10.636-027 SRC: Serial
17:02:10.638-027 CMD: Grp 0, Cmd 'SERIALLOG', Idx 1, Len 1, Pld 3, Data '3'
17:02:10.646-024 MQT: stat/wemos4/RESULT = {"SerialLog":{"3":{"Active":"3"}}}
17:02:11.601-027 CFG: Saved to flash at F8, Count 704, Bytes 4096
```

**Reply by arendst** [reply]
Latest changes now also support AIM ID. Enable it using command ``GM861Zone208 0x80``.

Expect now outputs like:
```
17:23:02.587-024 MQT: tele/wemos4/SENSOR = {"Time":"2023-05-06T17:23:02","GM861":{"AIM":"E0","Code":"8711218972875"}}
```
with AIM ID enabled and:
```
17:23:02.587-024 MQT: tele/wemos4/SENSOR = {"Time":"2023-05-06T17:23:02","GM861":{"Code":"8711218972875"}}
```
with AIM ID disabled. 
Notice the change from now obsolete:
```
17:00:05.003-024 MQT: tele/wemos4/SENSOR = {"Time":"2023-05-05T17:00:04","GM861":"8711218972875"}
```

**Reply by arendst** [reply]
NB. See head of file ``xsns_107_gm861.ino`` for latest command syntax

**Reply by arendst** [reply]
And now also optionally decodes the AIM-Id when `#define GM861_DECODE_AIM` is used:
```
16:11:23.948-024 MQT: tele/wemos4/SENSOR = {"Time":"2023-05-07T16:11:23","GM861":{"AIM":"EAN13","Code":"8711218972875"}}
16:12:16.339-024 MQT: tele/wemos4/SENSOR = {"Time":"2023-05-07T16:12:16","GM861":{"AIM":"QRCode","Code":"https://bucket-refoss-static.refoss.net/refoss/production/staticfile/file/2023/03/16/16789591212549840.pdf"}}
```

**Comment by spoeh-man** [comment]
it Works also with GM 60 tested today. Only for Information.

**Reply by arendst** [reply]
Thx for reporting

**Comment by DutchCoban** [comment]
Thanks for the great work, with Tasmota and the GM861 driver.
In Home Assistant I don't see the GM861 as a sensor, other devices like GPS, BH1750, BME280 do.
Is it something I missed or hasn't that been implemented?

**Reply by sfromis** [reply]
Without having tried it, I'd not expect such a device to be automatically configured in HA, meaning that you likely need a manual configuration.

**Comment by Dtrieb** [comment]
FYI

[GM861.pdf](https://github.com/arendst/Tasmota/files/14449095/GM861.pdf)

**Comment by V69dxz** [comment]
Hi all,

I bought the GM60 like @spoeh-man ,  to create a barcode scanner connecting Grocy and HA [1] built by @ XDROLLOXD. 

Unfortunately, I see that the scanner recognizes barcodes, but does not transmit any to my ESP32 wroom (arduino serial monitor shows `null`). 

Moreover, I would like to test if it works correctly with Tasmota and a Wemos D1 (and my esp32). Can you confirm that I need GPIO 1 & 3 for the TX and RX cables @spoeh-man?

I have to add that this is completely new territory for me. 

thanks!

Ps. I also tried to run the example libraries posted on the GM60 github pages, without any success, but based on my interpretation of your discussion and the datasheet/manual, it can be that this is caused by having different baudrates? So I will check for this too.

[1] https://github.com/XDROLLOXD/ESP32_Grocy_Barcode_Scanner

**Reply by sfromis** [reply]
As a starting point, connect TX (transmit) from the reader to RX (receive) on the Tasmota end, and vice versa. And yes, with ESP8266 ESP TX is gpio 1 and RX is gpio 3. With ESP32, you pick the gpio pins used yourself.

**Reply by V69dxz** [reply]
Thanks! Ah 👍 will do!

**Reply by V69dxz** [reply]
Can you verify, the following:
- I flashed the esp32 (az delivery) with tasmoto
- In the GPIO configuration I set pin 25 to "serial TX" and pin 26 to "serial RC" (and attach the RX and TX wires to those pins)
- In the console I run the command `#define USE_GM861`, including the hash #
- If everything should work, I would see the results of the barcode scanner in the console?

If that's correct, than I have a magic issue with my wires...

Nice to learn Tasmota b2w, really cool :)

**Reply by barbudor** [reply]
> In the console I run the command #define USE_GM861, including the hash #

That will have no other effect than generating an error

You are confusing with user_config_override.h for the custom build

**Reply by barbudor** [reply]
So if you haven't got that yet: the driver for that is not included in any default tasmota binary pre-built
You have to compile your own binary with one of the supported and documented methods

**Reply by Dtrieb** [reply]
Are the driver really not in the last firmware?
@arendst

It I look in the docu, it's listed there.

**Reply by sfromis** [reply]
It is [listed as not being included in any of tht precompiled builds](https://tasmota.github.io/docs/BUILDS/). You need to do a [build](https://tasmota.github.io/docs/Compile-your-build/) to have it included.

**Comment by V69dxz** [comment]
@barbudor , ok. All works well for me… the thing was that the barcode scanner needs a faiir amount of light to work… with hindsign I started to test the scanner in a dimmed room, then nothing worked and started to do a kind of russian roulette with the wires… 


Thanks for the help!

**Comment by steeleto** [comment]
Subject: Help Needed with GM60 Scanner Module & ESP32-S3 Integration

Hi everyone,

I'm having trouble getting the GM60 scanner module to work with my ESP32-S3. Here’s what I've done so far:

Hardware Setup: I connected the RX and TX cables of the GM60 to the ESP32-S3’s hardware Serial1 pins (configured on GPIO 15 and 16 based on the datasheet).

Software Testing: At first, I was able to receive decoded data, but when I try to troubleshoot or adjust settings, the data stops coming through entirely. To avoid re-wiring every time, I change the defined pins in my program—which seems like a faster option—but it hasn’t resolved the intermittent issues.

I have a couple of questions:

Configuration Method: Do I need to set the GM60 scanner’s functions by writing zone commands for each function, or is it sufficient to enable hardware serial communication and configure it by scanning the 2D codes mentioned in the manual?

Driver Availability: Why isn’t there a unified driver provided? With over 20 scanner modules available globally, wouldn’t it be much simpler to have one driver where you can just comment or uncomment the relevant lines for each module?

Any assistance or insights would be greatly appreciated as I need to move forward with my project.

Thanks in advance!

**Reply by sfromis** [reply]
Sure, someone writing an unified driver for "over 20 scanner modules", and submitting a PR, does sound nice. Who's going to do it?

BTW. There is no such thing as specific hardware serial pins for ESP32, Tasmota configuration allows you to select which gpios, which then becomes hardware serial.

With no details, no basis for even guessing at what could be behind your issues.

**Comment by steeleto** [comment]
yes I set on the esp32 s3 the recommended pins 15,16 and set to  hardware serial 1 , I have write even the most zones to reset , command and save the commands on the gm 60 . Still nothing , all I want is to receive the barcode number from the barcodes I scan and printed as strings  no matter what type of barcode  is . Is probably the DFRobot_Gm_60 driver  working for any type of GM_60 scanner module ?

**Reply by sfromis** [reply]
Are you even using Tasmota? If not, don't expect answers here. If you are, you still posted no specific details of configuration or testing.

Also, Tasmota does not even have a driver for GM60, meaning that only chance would be if you you can handle it yourself via some sort of generic serial support, like with Berry. Or if "someone" were to create a driver.

**Comment by icikadha03-cpu** [comment]
Can someone help me connect the data from this GM861 to an ESP32? It’s really confusing and frustrating!

**Reply by sfromis** [reply]
You need a Tasmota custom build with `#define USE_GM861`, and then wiring and configuring gpios for `GM861 Rx` and `GM861 Rx`. Of course also connect VCC and GND.
